dispatcher/MongoConnect.py

```diff
@@ -261,66 +269,122 @@ def aggregate_status(self):
                     doc = self.collections['incoming_commands'].find_one(
                             {'key': f'{detector}.{key}'}, sort=[('_id', -1)])
                     if doc is None:
-                        self.log.error('No key %s for %s???' % (key, detector))
+                        self.log.error(f'No key {key} for {detector}???')
                         return None
                     latest_settings[detector][doc['field']] = doc['value']
                     if latest is None or doc['time'] > latest:
                         latest = doc['time']
                         latest_settings[detector]['user'] = doc['user']
-            self.latest_settings = latest_settings
-            return self.latest_settings
+            self.goal_state = latest_settings
+            return self.goal_state
         except Exception as e:
             self.log.debug(f'get_wanted_state failed due to {type(e)} {e}')
             return None
 
-    def get_configured_nodes(self, detector, link_mv, link_nv):
+    def is_linked(self, a, b):
+        '''
+        Check if the detectors are in a compatible linked configuration.
         '''
-        Get the nodes we want from the config file
+        mode_a = self.goal_state[a]["mode"]
+        mode_b = self.goal_state[b]["mode"]
+        doc_a = self.collections['options'].find_one({'name': mode_a})
+        doc_b = self.collections['options'].find_one({'name': mode_b})
+        detectors_a = doc_a['detector']
+        detectors_b = doc_b['detector']
+
+        # Check if the linked detectors share the same run mode and
+        # if they are both present in the detectors list of that mode
+        return mode_a == mode_b and a in detectors_b and b in detectors_a
+
+    def get_super_detector(self):
         '''
-        retnodes = []
-        retcc = []
-        retnodes = list(self.latest_status[detector]['readers'].keys())
-        retcc = list(self.latest_status[detector]['controller'].keys())
-        if detector == 'tpc' and link_nv == 'true':
-            retnodes += list(self.latest_status['neutron_veto']['readers'].keys())
-            retcc += list(self.latest_status['neutron_veto']['controllers'].keys())
-        if detector == 'tpc' and link_mv == 'true':
-            retnodes += list(self.latest_status['muon_veto']['readers'].keys())
-            retcc += list(self.latest_status['muon_veto']['controllers'].keys())
-        return retnodes, retcc
+        Get the Super Detector configuration
+        if the detectors are in a compatible linked mode.
+        - case A: tpc, mv and nv all linked
+        - case B: tpc, mv and nv all un-linked
+        - case C: tpc and mv linked, nv un-linked
+        - case D: tpc and nv linked, mv un-linked
+        - case E: tpc unlinked, mv and nv linked
+        We will check the compatibility of the linked mode for a pair of detectors per time.
+        '''
+        ret = {'tpc': {'controller': list(self.dc['tpc']['controller'].keys()),
```

Comment:
I'm not fully convinced that this is cleaner/more robust than having a `ret = {'linked' : ...}` document, especially with the hypervisor having only the power over a third of the system I'm worried things will break.

Reply:
If we change the format of this dict then we have to completely rewrite the rest of the package, because that's designed to loop over whichever logical detectors are determined by this function.